File: jar/src/main/java/jline/api/mam/Qbd_raprap1.java
/**
 * @file Quasi-Birth-Death process RAP/RAP/1 queue analysis
 *
 * <p>The two RAPs are INDEPENDENT of each other, so the QBD phase space is the
 * product of the two phase spaces and the blocks factor as Kronecker products.
 * This class is the thin product-space wrapper; the analysis itself is the
 * block-level core {@link Qbd_rap}, which solves an arbitrary QBD with RAP
 * components. A model whose arrival process and sequence of service times share
 * a phase space, and are therefore cross-correlated, has no such product
 * structure and must call {@link Qbd_rap} directly.</p>
 *
 * <p>References: N. G. Bean and B. F. Nielsen, "Quasi-Birth-and-Death Processes
 * with Rational Arrival Process Components", Stochastic Models, 26(3), 2010,
 * pp. 309-334. The analysis rests on the prediction-process interpretation of a
 * RAP due to Asmussen and Bladt, which is what allows a QBD argument to be
 * carried over to matrices that are not nonnegative. The same prediction
 * process underlies the conditional-vector RAP sampler in {@link Rap_sample}.</p>
 *
 * <p>Phase ordering: the QBD phase is the pair (arrival phase, service phase)
 * laid out with the ARRIVAL phase major and the service phase minor, i.e. the
 * phase index is a*ns + s. That is the ordering produced by
 * {@code RAPa.kron(I_ns)} and {@code I_na.kron(RAPs)}, and pqueue is indexed by
 * it downstream, so the two Kronecker factors must not be swapped.</p>
 *
 * @since LINE 3.0
 */
package jline.api.mam;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Qbd_raprap1 {
    private Qbd_raprap1() {}

    /**
     * Cap on the number of reported level vectors. The level series is cut at
     * the same point in MATLAB, the JAR and Python, otherwise the mean queue
     * lengths disagree in the truncated tail.
     */
    private static final int MAX_NUM_COMP = 100;

    /** Mass threshold at which the level series is truncated. */
    private static final double MASS_TOL = 1e-10;

    public static QbdRapRap1Result qbd_raprap1(MatrixCell RAPa, MatrixCell RAPs, Double util) {
        int na = RAPa.get(0).getNumRows();
        int ns = RAPs.get(0).getNumRows();

        MatrixCell scaledRAPs = RAPs;
        if (util != null) {
            double lambdaScale = Map_lambda.map_lambda(RAPa);
            scaledRAPs = Map_scale.map_scale(RAPs, util.doubleValue() / lambdaScale);
        }

        double lambdaA = Map_lambda.map_lambda(RAPa);

        // QBD blocks of the RAP/RAP/1 queue: a level is the number in system, a
        // phase is the (arrival,service) RAP phase pair, arrival phase major.
        // The level rises on an arrival (F), falls on a service completion (B),
        // and the two RAPs evolve independently between level changes (L, the
        // Kronecker sum of the two hidden generators). At level 0 the queue is
        // empty, so no service completion can occur and only the arrival RAP
        // evolves: the boundary local block is B1 = kron(Ca,I) and the boundary
        // up-block is F.
        Matrix IA = Matrix.eye(na);
        Matrix IS = Matrix.eye(ns);
        Matrix F = RAPa.get(1).kron(IS);
        Matrix L = RAPa.get(0).kron(IS).add(IA.kron(scaledRAPs.get(0)));
        Matrix B = IA.kron(scaledRAPs.get(1));
        Matrix B1 = RAPa.get(0).kron(IS);

        // Theorem 7 of Bean and Nielsen (2010): G from the quadratic matrix
        // equation, U = L + F*G, R = F*inv(-U), and pi0 from the boundary
        // equation pi0*(B1 + R*B) = 0 normalised by pi0*inv(I-R)*e = 1.
        QbdRapResult core = Qbd_rap.qbd_rap(F, L, B, F, B1, 0);
        Matrix R = core.getR();
        Matrix G = core.getG();
        Matrix pi0 = core.getPi0();

        // Level series, truncated when the accumulated mass reaches 1-1e-10 or
        // at MAX_NUM_COMP level vectors. Qbd_rap returns the exact mean queue
        // length in closed form, but this method keeps reporting the truncated
        // series because pqueue is the documented return value and the three
        // codebases must cut the tail at exactly the same point.
        int n = na * ns;
        java.util.List<Matrix> levels = new java.util.ArrayList<Matrix>();
        levels.add(pi0);
        double sumpi = rowSum(pi0);
        int numit = 1;
        while (sumpi < 1.0 - MASS_TOL && numit < 1 + MAX_NUM_COMP) {
            Matrix next = levels.get(numit - 1).mult(R);
            levels.add(next);
            numit++;
            sumpi += rowSum(next);
        }

        int numLevels = levels.size();
        Matrix pqueue = new Matrix(numLevels, n);
        for (int i = 0; i < numLevels; i++) {
            Matrix lvl = levels.get(i);
            for (int j = 0; j < n; j++) {
                pqueue.set(i, j, lvl.get(0, j));
            }
        }

        double etaVal = core.getSpr();
        Matrix eta = new Matrix(1, 1);
        eta.set(0, 0, etaVal);

        double XN = lambdaA;
        double UN;
        double QN;

        if (na == 1 && ns == 1) {
            UN = 1.0 - pqueue.get(0, 0);
        } else {
            double sum0 = 0.0;
            for (int j = 0; j < n; j++) {
                sum0 += pqueue.get(0, j);
            }
            UN = 1.0 - sum0;
        }

        QN = 0.0;
        for (int i = 0; i < numLevels; i++) {
            double levelProb = 0.0;
            for (int j = 0; j < n; j++) {
                levelProb += pqueue.get(i, j);
            }
            QN += (double) i * levelProb;
        }

        return new QbdRapRap1Result(XN, QN, UN, pqueue, R, eta, G, B, L, F);
    }

    public static QbdRapRap1Result qbd_raprap1(MatrixCell RAPa, MatrixCell RAPs) {
        return qbd_raprap1(RAPa, RAPs, null);
    }

    private static double rowSum(Matrix row) {
        double s = 0.0;
        for (int j = 0; j < row.getNumCols(); j++) {
            s += row.get(0, j);
        }
        return s;
    }
}

File: jar/src/main/kotlin/jline/api/mam/Qbd_raprap1.kt
/**
 * @file Quasi-Birth-Death process RAP/RAP/1 queue analysis
 *
 * Analyzes RAP/RAP/1 queueing systems using QBD methods with rational arrival processes.
 * Extends QBD analysis to more general arrival and service process models.
 *
 * @since LINE 3.0
 */
package jline.api.mam

import jline.lib.smc.QBD_CR
import jline.lib.smc.QBD_pi
import jline.util.matrix.Matrix
import jline.util.matrix.MatrixCell
import org.ejml.dense.row.factory.DecompositionFactory_DDRM
import org.ejml.data.DMatrixRMaj

/**
 * Analyze RAP/RAP/1 queue using QBD methods
 *
 * @param RAPa Arrival RAP (Rational Arrival Process)
 * @param RAPs Service RAP (Rational Service Process)
 * @param util Optional utilization parameter
 * @return QbdRapRap1Result containing performance measures
 */
data class QbdRapRap1Result(
    val XN: Double,        // Throughput
    val QN: Double,        // Queue length
    val UN: Double,        // Utilization
    val pqueue: Matrix,    // Queue state probabilities
    val R: Matrix,         // Rate matrix
    val eta: Matrix,       // Caudal characteristic (scalar stored as 1x1)
    val G: Matrix,         // G matrix
    val B: Matrix,         // Backward matrix
    val L: Matrix,         // Local matrix
    val F: Matrix          // Forward matrix
)

fun qbd_raprap1(RAPa: MatrixCell, RAPs: MatrixCell, util: Double? = null): QbdRapRap1Result {
    val na = RAPa[0].numRows
    val ns = RAPs[0].numRows

    var scaledRAPs = RAPs
    if (util != null) {
        val lambdaA = map_lambda(RAPa)
        scaledRAPs = map_scale(RAPs, util / lambdaA)
    }

    val lambdaA = map_lambda(RAPa)
    val lambdaS = map_lambda(scaledRAPs)
    val actualUtil = lambdaA / lambdaS

    // Construct QBD matrices for RAP/RAP/1 using Kronecker sum for L
    val IA = Matrix.eye(na)
    val IS = Matrix.eye(ns)
    val F = RAPa[1].kron(IS)
    val L = RAPa[0].kron(IS).add(IA.kron(scaledRAPs[0]))  // Kronecker sum
    val B = IA.kron(scaledRAPs[1])

    // Boundary matrices
    val B0 = B.copy()
    val B1 = RAPa[0].kron(IS)

    // Solve using QBD CR with RAP components (RAPComp=1)
    val qbdResult = QBD_CR(B, L, F, null, null, null, 1)
    val R = qbdResult["R"] ?: throw RuntimeException("QBD_CR failed to compute R matrix")
    val G = qbdResult["G"] ?: throw RuntimeException("QBD_CR failed to compute G matrix")

    // Compute stationary distribution using QBD_pi with RAP components
    val pi = QBD_pi(B0, B1, R, 100, 0, null, 1)
    val n = na * ns
    val numLevels = pi.numCols / n

    // Reshape pqueue: each row is a level's distribution summed over phases
    val pqueue = Matrix(numLevels, n)
    for (i in 0 until numLevels) {
        for (j in 0 until n) {
            pqueue[i, j] = pi[0, i * n + j]
        }
    }

    // Compute eta = max(abs(eig(R)))
    val etaVal = spectralRadius(R)
    val eta = Matrix(1, 1)
    eta[0, 0] = etaVal

    // Compute performance measures
    val XN = lambdaA
    var UN: Double
    var QN: Double

    if (na == 1 && ns == 1) {
        UN = 1.0 - pqueue[0, 0]
    } else {
        var sum0 = 0.0
        for (j in 0 until n) {
            sum0 += pqueue[0, j]
        }
        UN = 1.0 - sum0
    }

    // QN = sum over levels k of k * sum_j(pqueue(k,j))
    QN = 0.0
    for (i in 0 until numLevels) {
        var levelProb = 0.0
        for (j in 0 until n) {
            levelProb += pqueue[i, j]
        }
        QN += i.toDouble() * levelProb
    }

    return QbdRapRap1Result(XN, QN, UN, pqueue, R, eta, G, B, L, F)
}

/**
 * Compute spectral radius (maximum absolute eigenvalue) of a matrix
 */
private fun spectralRadius(A: Matrix): Double {
    val n = A.numRows
    val dm = DMatrixRMaj(n, n)
    for (i in 0 until n) {
        for (j in 0 until n) {
            dm[i, j] = A[i, j]
        }
    }
    val evd = DecompositionFactory_DDRM.eig(n, false)
    evd.decompose(dm)
    var maxAbs = 0.0
    for (i in 0 until evd.numberOfEigenvalues) {
        val ev = evd.getEigenvalue(i)
        val absVal = Math.sqrt(ev.real * ev.real + ev.imaginary * ev.imaginary)
        if (absVal > maxAbs) {
            maxAbs = absVal
        }
    }
    return maxAbs
}

/**
 * QBD raprap1 algorithms
 */
@Suppress("unused")
class QbdRaprap1 {
    companion object {
        // Class documentation marker for Dokka
    }
}
